doc/src/manual/code-loading.md

```diff
@@ -0,0 +1,324 @@
+
+Since the decentralized naming problem is somewhat abstract, it may help to walk through a concrete scenario to understand the issue. Suppose you're developing an application called `App`, which uses two packages: `Pub` and  `Priv`. `Priv` is a private package that you created, whereas `Pub` is a public package that you use but don't control. When you created `Priv`, there was no public package by that name. Subsequently, however, an unrelated package also named `Priv` has been published and become popular. In fact, the `Pub` package has started to use it. Therefore, when you next upgrade `Pub` to get the latest bug fixes and features, `App` will end up—through no action of yours other than upgrading—depending on two different packages named `Priv`. `App` has a direct dependency on your private `Priv` package, and an indirect dependency, through `Pub`, on the new public `Priv` package. Since these two `Priv` packages are different but both required for `App` to continue working correctly, the expression `import Priv` must refer to different `Priv` packages depending on whether it occurs in `App`'s code or in `Pub`'s code. Julia's package loading mechanism allows this by distinguishing the two `Priv` packages by context and UUID. How this distinction works is determined by environments, as explained in the following sections.
+
+## Environments
+
+An *environment* determines what `import X` and `using X` mean in various code contexts and what files these statements cause to be loaded. Julia understands three kinds of environments:
+
+1. **A project environment** is a directory with a project file and an optional manifest file. The project file determines what the names and identities of the direct dependencies of a project are. The manifest file, if present, gives a complete dependency graph, including all direct and indirect dependencies, exact versions of each dependency, and sufficient information to locate and load the correct version.
+2. **A package directory** is a directory containing the source trees of a set of packages as subdirectories. This kind of environment was the only kind that existed in Julia 0.6 and earlier. If `X` is a subdirectory of a package directory and `X/src/X.jl` exists, then the package `X` is available in the package directory environment and `X/src/X.jl` is the source file by which it is loaded.
```

Comment:
Would be good to mention that packages also have project files. And, does that mean it makes sense to say that a package is also a project environment?

Reply:
Each package with a project file does have an associated environment, yes. I feel like that might just be confusing things here though since that environment doesn't really come into play here.

Reply:
I think it would be good to point out a clarification that you made someday on Slack - that having a manifest file would allow one to replicate the exact state of the software at some future time thus ensuring replicability, which is what Julia's aim is about. (This was the discussion with David Anthoff IIRC.) 
Having that cleared up after the bullet points where you introduce the manifest file would make its existence cogent and clear.

Reply:
Ok, good idea. The documentation of this is a bit fractured because half of it—the part about how packages get loaded—belongs in the language manual since it's built into the language itself, while the other half—the big about how to use Pkg3 to create and manipulate project and manifest files doesn't since it's an external tool. But a little bit of repetition won't hurt.